File: command-queue-sample/src/main/java/com/zhuinden/commandqueuesample/MainActivity.java
package com.zhuinden.commandqueuesample;

import android.annotation.SuppressLint;
import android.content.Context;
import android.os.Bundle;
import android.support.annotation.NonNull;
import android.support.design.widget.BottomNavigationView;
import android.support.v7.app.AppCompatActivity;
import android.util.Log;
import android.view.ViewGroup;

import java.util.Arrays;

import butterknife.BindView;
import butterknife.ButterKnife;

public class MainActivity
        extends AppCompatActivity {
    private static final String TAG = "MainActivity";

    @BindView(R.id.navigation)
    BottomNavigationView navigation;

    @BindView(R.id.root)
    ViewGroup root;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        setContentView(R.layout.activity_main);
        ButterKnife.bind(this);

        navigation.setOnNavigationItemSelectedListener(item -> {
            switch(item.getItemId()) {
                case R.id.navigation_home:

                    return true;
                case R.id.navigation_dashboard:

                    return true;
                case R.id.navigation_notifications:

                    return true;
            }
            return false;
        });
    }
}

File: command-queue-sample/src/main/java/com/zhuinden/commandqueuesample/MainActivity.kt
package com.zhuinden.commandqueuesample

import android.os.Bundle
import android.view.Window
import android.widget.Toast
import androidx.appcompat.app.AppCompatActivity
import androidx.lifecycle.Observer
import com.zhuinden.commandqueuesample.databinding.ActivityMainBinding

class MainActivity : AppCompatActivity() {
    private val viewModel by lazy { createViewModel { MainViewModel() } }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)

        setContentView(R.layout.activity_main)

        val binding = ActivityMainBinding.bind(findViewById(Window.ID_ANDROID_CONTENT))

        binding.button.onClick {
            viewModel.doSomethingInteresting()
        }

        viewModel.loading.observe(this, Observer { loading: Boolean? ->
            binding.loadingOverlay.showIf { loading!! }
        })
    }

    override fun onStart() {
        super.onStart()
        viewModel.commandQueue.setReceiver { command ->
            when (command) {
                is MainViewModel.Events.DoSomething -> showToast("Do something!", Toast.LENGTH_SHORT)
                is MainViewModel.Events.DoOtherThing -> showToast("Do other thing!", Toast.LENGTH_SHORT)
            }.safe()
        }
    }

    override fun onStop() {
        viewModel.commandQueue.detachReceiver()
        super.onStop()
    }
}

